File: compiler/bench/power.c
// expected: 1073741824
#include <stdio.h>
long power(long base, long exp) {
    if (exp == 0) return 1;
    if (exp & 1) {
        return base * power(base, exp - 1);
    } else {
        long half = power(base, exp / 2);
        return half * half;
    }
}

int main() {
    for (long i = 0; i < 10000000; i++) {
        power(2, 30);
    }
    printf("%ld\n", power(2, 30));
    return 0;
}

File: compiler/bench/power.fs
\ expected: 1073741824
\ Recursive power, 10M calls

: power ( base exp -- result ) recursive
  dup 0= if 2drop 1 exit then
  dup 1 and if
    1- over swap recurse *
  else
    2/ over swap recurse dup *
  then nip ;

: main
  10000000 0 do 2 30 power drop loop
  2 30 power . cr ;
